Mathlib/RingTheory/HopfAlgebra.lean

```diff
@@ -115,3 +125,152 @@ instance toHopfAlgebra : HopfAlgebra R R where
 theorem antipode_eq_id : antipode (R := R) (A := R) = .id := rfl
 
 end CommSemiring
+namespace MonoidAlgebra
+
+variable {R : Type u} {A : Type v} [CommSemiring R] [Semiring A] [HopfAlgebra R A]
+variable {G : Type w} [Group G]
+
+variable (R A G) in
+/-- The antipode in the `R`-Hopf algebra structure on `MonoidAlgebra A G`, sending
+`single g a` to `single g⁻¹ (antipode a)`. -/
+def antipode :
+    MonoidAlgebra A G →ₗ[R] MonoidAlgebra A G :=
+  Finsupp.lsum R fun g => Finsupp.lsingle g⁻¹ ∘ₗ HopfAlgebra.antipode
```

Comment:
This definition is arguably a bad idea, because now you have separate `HopfAlgebra.antipode` and `MonoidAlgebra.algebra` definitions, and lemmas have to pick which spelling to use.

Some options here:
* Make it clear in the docstring that `HopfAlgebra.antipode` is the preferred spelling
* Add a `simp` lemma turning this spelling into `HopfAlgebra.antipode`
* Pull out a `HopfAlgebraStruct` base class from `HopfAlgebra` that contains just the antipode (and perhaps bialgebra structure), so that you can create this data first before proving `antipode_single`.

Reply:
Thanks for the review! I like the third option so I've gone with that.